tensorflow_datasets/video/moving_sequence.py

```diff
@@ -0,0 +1,419 @@
+      on [0, 1].
+    x0: [n_trajectories, ndims] `dtype` tensor of random initial positions
+      used
+    velocity: [n_trajectories, ndims] `dtype` tensor of random normalized
+      velocities used.
+  """
+  if not dtype.is_floating:
+    raise ValueError("dtype must be float")
+  speed = tf.convert_to_tensor(speed, dtype=dtype)
+  if speed.shape.ndims not in (0, 1):
+    raise ValueError("speed must be scalar or rank 1 tensor")
+
+  nt = n_trajectories
+  x0 = tf.random.uniform((nt, ndims), dtype=dtype)
+  velocity = tf.random_normal((nt, ndims), dtype=dtype)
+  speed = tf.convert_to_tensor(speed, dtype=dtype)
+  if speed.shape.ndims == 1:
+      if speed.shape[0].value not in (1, n_trajectories):
+        raise ValueError(
+            "If speed is a rank 1 tensor, its length must be 1 or same as "
+            "`n_trajectories`, got shape %s" % str(speed.shape))
+      speed = tf.expand_dims(speed, axis=-1)
+  velocity = velocity * (
+    speed / tf.linalg.norm(velocity, axis=-1, keepdims=True))
+  t = tf.range(sequence_length, dtype=dtype)
+  linear_trajectories = get_linear_trajectories(x0, velocity, t)
+  bounced_trajectories = bounce_to_bbox(linear_trajectories)
+  return bounced_trajectories, x0, velocity
+
+
+def get_linear_trajectories(x0, velocity, t):
+  """
+  Args:
+    x0: [n_trajectories, ndims] float tensor.
+    velocity: [n_trajectories, ndims] float tensor
+    t: [sequence_length] float tensor
+
+  Returns:
+    x: [n_trajectories, sequence_length, ndims] float tensor.
+  """
+  x0 = tf.convert_to_tensor(x0)
+  velocity = tf.convert_to_tensor(velocity)
+  if x0.shape.ndims != 2:
+    raise ValueError("x0 must be a rank 2 tensor")
+  if velocity.shape.ndims != 2:
+    raise ValueError("velocity must be a rank 2 tensor")
+  if t.shape.ndims != 1:
+    raise ValueError("t must be a rank 1 tensor")
+  x0 = tf.expand_dims(x0, axis=1)
+  velocity = tf.expand_dims(velocity, axis=1)
+  dx = velocity * tf.expand_dims(tf.expand_dims(t, axis=0), axis=-1)
+  linear_trajectories = x0 + dx
+  assert linear_trajectories.shape.ndims == 3, \
+    "linear_trajectories should be a rank 3 tensor"
+  return linear_trajectories
+
+
+def bounce_to_bbox(points):
+  """
+  Bounce potentially unbounded points to [0, 1].
+
+  Bouncing occurs by exact reflection, i.e. a pre-bound point at 1.1 is moved
+  to 0.9, -0.2 -> 0.2. This theoretically can occur multiple times, e.g.
+  2.3 -> -0.7 -> 0.3
+
+  Implementation
+  points <- points % 2
+  return min(2 - points, points)
+
+  Args:
+    points: float array
+
+  Returns:
+    tensor with same shape/dtype but values in [0, 1].
+  """
+  points = points % 2
+  return tf.math.minimum(2 - points, points)
+
+
+MovingSequence = collections.namedtuple(
+  "MovingSequence",
+  ["image_sequence", "trajectories", "start_positions", "velocities"])
+
+
+def images_to_moving_sequence(
+    images, sequence_length=20, speed=0.1, total_padding=(36, 36),
+    **kwargs):
```

Comment:
Since this is the main function, let’s not have kwargs and instead list out all the arguments.